### PR TITLE
Add Semi-SL for Multi-class Classification (EfficientNet-B0)

### PR DESCRIPTION
**Pull request by harimkang**
### Summary

CVS-141546
- OTX 2.1 imports the Semi-SL algorithm from OTX 1.6.
- Add Unlabeled data pipeline
```shell
--data_root: Target the supervised folder.
--data.config.unlabeled_subset.data_root: Targets the unlabeled folder.
```

image
For the example above, 
`--data_root` is `supervised/`
and `--data.config.unlabeled_subset.data_root` must be specified as `unlabeled/`.

- Add Multi-Transforms logic into unlabeled dataset
image
- Add Semi-SL classifier & Head in OTX 1.6

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>

<div style='direction:ltr'>

FMNIST   25@1_0 | Accuracy | E2E   time | Epochs
-- | -- | -- | --
Baseline   (SL) | 79.01 | 0:03:37 | 90
Semi-SL | 83.03     (+4.02) | 0:06:19   (1.75x) | 75

</div>

</body>

</html>

### How to test

```shell
otx train --config src/otx/recipe/classification/multi_class_cls/efficientnet_b0_semisl.yaml --data_root fmnist@25_0/supervised --data.config.unlabeled_subset.data_root fmnist@25_0/unlabeled
```

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by eunwoosh — DISMISSED** [review]

**Review by eunwoosh — DISMISSED** [review]

File: src/otx/algo/classification/efficientnet.py [hunk 1]
```diff
@@ -172,6 +178,93 @@ def forward_for_tracing(self, image: Tensor) -> Tensor | dict[str, Tensor]:
...
+            dict[str, Any]: The customized input data.
+        """
+        if self.training:
+            mode = "loss"
+        elif self.explain_mode:
+            mode = "explain"
+        else:
+            mode = "predict"
+
+        if isinstance(inputs, dict):
```
Comment (kprokofi) [inline src/otx/algo/classification/efficientnet.py:221]: Do we have cases, when inputs come not as a dict in SemiSL mode?

File: src/otx/core/config/data.py [hunk 2]
```diff
@@ -90,6 +90,24 @@ class VisualPromptingConfig:
...
 
+@dataclass
+class UnlabeledDataConfig(SubsetConfig):
+    """DTO for unlabeled data."""
+
+    data_root: str | None = None
+    data_format: str = "image_dir"
+
+    batch_size: int = 0
+    subset_name: str = "unlabeled"
```
Comment (kprokofi) [inline src/otx/core/config/data.py:101]: Do we need subset name in config since it is always should be unlabeled?

File: src/otx/core/data/module.py [hunk 3]
```diff
@@ -170,7 +217,18 @@ def train_dataloader(self) -> DataLoader:
...
-        return DataLoader(**common_args)
+        dataloader: DataLoader = DataLoader(**common_args)
+        if (unlabeled_dataloader := self.unlabeled_dataloader()) is not None:
+            # Utilize the CombinedLoader provided by Lightning to bundle multiple dataloaders.
+            # https://lightning.ai/docs/pytorch/stable/data/iterables.html
+            from lightning.pytorch.utilities import CombinedLoader
+
+            iterables = {
+                "labeled": dataloader,
+                **unlabeled_dataloader,
```
Comment (kprokofi) [inline src/otx/core/data/module.py:241]: Why not just "unlabeled" : unlabeled_dataloader ?

File: src/otx/core/data/module.py [hunk 4]
```diff
@@ -217,6 +275,63 @@ def predict_dataloader(self) -> DataLoader:
...
+            "batch_size": config.batch_size,
+            "num_workers": config.num_workers,
+            "pin_memory": True,
+            "persistent_workers": config.num_workers > 0,
+        }
+
+        dataloaders = {}
+        if isinstance(config.transforms, dict):
+            log.warning(f"Unlabeled dataset has multiple transforms : {list(config.transforms.keys())}")
+            common_args["worker_init_fn"] = lambda _: torch.manual_seed(0)  # type: ignore[assignment]
```
Comment (kprokofi) [inline src/otx/core/data/module.py:315]: Is it OK to hardcode the seed? Do we need to change worker_init_fn?

File: src/otx/core/data/module.py [hunk 5]
```diff
@@ -217,6 +275,63 @@ def predict_dataloader(self) -> DataLoader:
...
+            for key in config.transforms:
+                dataset = self._get_dataset(key)
+                # For unlabeled datasets using Multi-Transforms, must use generators and samplers
+                # with the same seed to get the same data.
+                generator = torch.Generator().manual_seed(0)
```
Comment (kprokofi) [inline src/otx/core/data/module.py:320]: same about seed here

**Review by kprokofi — COMMENTED** [review]
Thanks for your work! Overall design looks good.

**Comment by kprokofi** [comment]
Can we simplify --data.config.unlabeled_subset.data_root CLI command? I prefer shorter, maybe we can create a link to this param and call it as "--unlabeled_data_root"

File: src/otx/core/data/module.py (hunk 3 above)
Comment (harimkang) [reply src/otx/core/data/module.py:241]: In this implementation `unlabeled_dataloader` is a dict. This consists of the same dataset but each dataloader with multiple different transform types. So this dict can have multiple keys.

File: src/otx/core/data/module.py (hunk 4 above)
Comment (harimkang) [reply src/otx/core/data/module.py:315]: What I am considering here is the same dataset, but applying several different transforms. At this time, the unlabeled dataloader should return an index of the same dataset in the batch. So that code was added. Is there perhaps a better way?

File: src/otx/core/data/module.py (hunk 5 above)
Comment (harimkang) [reply src/otx/core/data/module.py:320]: This is to enable multiple unlabeled dataloaders to sample the same batch of data.

**Comment by harimkang** [comment]
> Can we simplify --data.config.unlabeled_subset.data_root CLI command? I prefer shorter, maybe we can create a link to this param and call it as "--unlabeled_data_root"

I think it's a good suggestion. I'll give it a try.

**Comment by harimkang** [comment]
> Can we simplify --data.config.unlabeled_subset.data_root CLI command? I prefer shorter, maybe we can create a link to this param and call it as "--unlabeled_data_root"

Looking at it, the current data-related configs are gathered in data.config. For now, I think we need to think a little more about how to make this a little shorter. Personally, I think it would be better if it goes along with the refactoring of `OTXDatamodule`.

File: src/otx/algo/classification/efficientnet.py (hunk 1 above)
Comment (harimkang) [reply src/otx/algo/classification/efficientnet.py:221]: My guess is that if the transforms are not dicts, we can receive unlabeled in the same way as other labeled subsets. I'll check this one more time.(use single transform case in unlabeled)

File: src/otx/core/config/data.py (hunk 2 above)
Comment (harimkang) [reply src/otx/core/config/data.py:101]: I'll remove this part. thank you!

Comment (harimkang) [reply src/otx/core/config/data.py:101]: I've been working on this, but there's a problem. Currently, `UnlabeledDataConfig` is a structure that inherits from `SubsetConfig`, so it needs that attribute. I'll leave it as it is for now.

File: src/otx/algo/callbacks/unlabeled_loss_warmup.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
+"""Module for defining hook for semi-supervised learning for classification task."""
+# Copyright (C) 2023 Intel Corporation
```
Comment (sungchul1) [inline src/otx/algo/callbacks/unlabeled_loss_warmup.py]:
```suggestion
# Copyright (C) 2024 Intel Corporation
```

**Review by sungchul1 — COMMENTED** [review]
LGTM

**Review by sungchul1 — DISMISSED** [review]

**Review by eunwoosh — DISMISSED** [review]

**Review by sungchul1 — APPROVED** [review]

**Review by eunwoosh — APPROVED** [review]
